upstairs/src/dummy_downstairs_tests.rs

```diff
@@ -156,102 +156,123 @@ pub(crate) mod protocol_test {
-            let packet = self
-                .fr
-                .lock()
-                .await
-                .next()
-                .await
-                .transpose()
-                .unwrap()
-                .unwrap();
-            match &packet {
-                Message::HereIAm {
-                    version,
-                    upstairs_id: _,
-                    session_id: _,
-                    gen: _,
-                    read_only: _,
-                    encrypted: _,
-                    alternate_versions: _,
-                } => {
-                    info!(self.inner.log, "negotiate packet {:?}", packet);
+            loop {
+                let packet = self
+                    .fr
+                    .lock()
+                    .await
+                    .next()
+                    .await
+                    .transpose()
+                    .unwrap()
+                    .unwrap();
+                match &packet {
+                    Message::HereIAm {
+                        version,
+                        upstairs_id: _,
+                        session_id: _,
+                        gen: _,
+                        read_only: _,
+                        encrypted: _,
+                        alternate_versions: _,
+                    } => {
+                        info!(self.inner.log, "negotiate packet {:?}", packet);
 
-                    self.fw
-                        .lock()
-                        .await
-                        .send(Message::YesItsMe {
-                            version: *version,
-                            repair_addr: self.inner.repair_addr,
-                        })
-                        .await
-                        .unwrap();
-                }
+                        self.fw
+                            .lock()
+                            .await
+                            .send(Message::YesItsMe {
+                                version: *version,
+                                repair_addr: self.inner.repair_addr,
+                            })
+                            .await
+                            .unwrap();
+                        break;
+                    }
 
-                _ => panic!("wrong packet"),
+                    Message::Ruok => {
+                        // A ping snuck in, ignore it
+                        continue;
+                    }
+                    x => panic!("wrong packet {:?}", x),
+                }
             }
 
-            let packet = self
-                .fr
-                .lock()
-                .await
-                .next()
-                .await
-                .transpose()
-                .unwrap()
-                .unwrap();
-            match &packet {
-                Message::PromoteToActive {
-                    upstairs_id,
-                    session_id,
-                    gen,
-                } => {
-                    assert!(*gen == 1);
+            loop {
```

Comment:
There should be a comment explaining the ping behaviour up here maybe, instead of in the match?

Reply:
Done